lib/spack/spack/ci.py

```diff
@@ -539,6 +577,97 @@ def _apply_section(dest, src):
                         job["attributes"] = self.__apply_submapping(
                             job["attributes"], job["spec"], section
                         )
+            elif has_dynmapping:
+                mapping = section["dynamic-mapping"]
+
+                # Get the endpoint
+                endpoint = mapping["endpoint"]
+                endpoint_url = urlparse(endpoint)
+
+                # Configure the request header
+                header = {"User-Agent": web_util.SPACK_USER_AGENT}
+                header.update(mapping.get("header", {}))
+
+                # Expand header environment variables
+                # ie. if tokens are passed
+                for value in header.values():
+                    value = os.path.expandvars(value)
+
+                verify_ssl = mapping.get("verify_ssl", spack.config.get("config:verify_ssl", True))
+                timeout = mapping.get("timeout", spack.config.get("config:connect_timeout", 10))
+
+                if "fields" in mapping:
+                    required = mapping["fields"].get("required", [])
+                    allowed = mapping["fields"].get("allow", [])
+                    ignored = mapping["fields"].get("ignore", [])
+                else:
+                    allowed = ["variables"]
+                    ignored = []
+                    required = []
+
+                # required keys are implicitly allowed
+                allowed = sorted(set(allowed + required))
+                ignored = sorted(set(ignored))
+                required = sorted(set(required))
+
+                def job_query(job):
+                    job_vars = job["attributes"]["variables"]
+                    query = {
+                        "spec": (
+                            "{SPACK_JOB_SPEC_PKG_NAME}@{SPACK_JOB_SPEC_PKG_VERSION}"
+                            # The preceding space is required (ref. https://github.com/spack/spack-gantry/blob/develop/docs/api.md#allocation)
+                            " {SPACK_JOB_SPEC_VARIANTS}"
+                            "%{SPACK_JOB_SPEC_COMPILER_NAME}@{SPACK_JOB_SPEC_COMPILER_VERSION}"
+                        ).format_map(job_vars)
+                    }
+                    return quote(query)
```

Comment:
```suggestion
                    query = (
                        "{SPACK_JOB_SPEC_PKG_NAME}@{SPACK_JOB_SPEC_PKG_VERSION}"
                        # The preceding space is required (ref. https://github.com/spack/spack-gantry/blob/develop/docs/api.md#allocation)
                        " {SPACK_JOB_SPEC_VARIANTS}"
                        "%{SPACK_JOB_SPEC_COMPILER_NAME}@{SPACK_JOB_SPEC_COMPILER_VERSION}"
                    ).format_map(job_vars)
                    return f"spec={quote(query)}"
```

I needed to make a couple tweaks to get it to work properly